### PR TITLE
feat(pipelines): add component_scope to internal metrics events

### PR DESCRIPTION
**Pull request by jdrouet**
Signed-off-by: Jérémie Drouet <jeremie.drouet@datadoghq.com>

### CODE REVIEW

File: src/topology/builder.rs
Comment (jszwedko) [inline src/topology/builder.rs:276]: How do these two differ? Could you given an example of both?

**Review by jszwedko — COMMENTED** [review]
Generally looks good to me. I just had a question about the difference between `id.to_string()` and `id.id().to_string()`.

**Comment by jszwedko** [comment]
Just a note that this is probably worth a release highlight too since it is adding a new label to metrics if we could add that as part of this PR (see `website/content/en/highlights`).

File: src/topology/builder.rs
Comment (jdrouet) [reply src/topology/builder.rs:276]: `id.to_string()` would give `pipeline_id#transform_name` but `id.id().to_string()` would just give `transform_name`.
That's why `component_scope` is added after to have the complete information.

Comment (jszwedko) [reply src/topology/builder.rs:276]: Ah, gotcha, I see. `component_name` is currently a deprecated tag that is always meant to match `component_id` to let users migrate over easily (as part of #8417). Given that, I think we should keep them matching here.

Comment (jdrouet) [reply src/topology/builder.rs:276]: I'm not sure I'd agree with that but I might be wrong. I'd say`component_name` is the behavior of having a single field whereas `component_id` would be the combination with `component_scope`. Therefore, we should have all the information in `component_name` that we have with `component_id` and `component_scope` combined 🤔

Comment (jszwedko) [reply src/topology/builder.rs:276]: 🤔 in #8417 we decided to just use `id` and deprecate `name`. `name` was left as a span field purely for backwards compatibility. One idea was that in the future users could perhaps specify human readable names as a `name = "Some Name"` config on components. We could decide to revive `name` with your definition, but I'm wondering if we could come up with a better term. Maybe `qualified_id`?

Comment (jdrouet) [reply src/topology/builder.rs:276]: 🤔 Ok. Maybe not to block this PR even more, I'll change what you asked for and we can see that afterward

**Comment by jszwedko** [comment]
@jdrouet it looks like there are still some outstanding comments here, but I'm happy to re-review once they are addressed.

**Comment by jdrouet** [comment]
@jszwedko Once #8810 is merged, I rebase all this, fix the issues and I'll ping you 😉 I hope it'll be ready before you appear in start working.

File: src/topology/builder.rs
Comment (spencergilbert) [inline src/topology/builder.rs:291]: `component_id.id()` still seems weird, I'm guessing there's a reason we can't do `component.id()`

File: src/topology/task.rs
Comment (spencergilbert) [inline src/topology/task.rs:63]: Similar to my other question

File: src/topology/builder.rs
Comment (jdrouet) [reply src/topology/builder.rs:291]: 🤔 I forgot this renaming from `component_id` to `component_key`...

File: src/topology/task.rs
Comment (jdrouet) [reply src/topology/task.rs:63]: Similar answer 😉

**Review by spencergilbert — APPROVED** [review]

**Comment by jdrouet** [comment]
@jszwedko feel free to merge it if it looks ok to you 😉

File: src/topology/builder.rs
Comment (lukesteensen) [inline src/topology/builder.rs:294]: I think this is my fault because I told you wrong last time, but `?` == `Debug` and `%` == `Display`, so I think we want `%` here and similar places 😢

**Review by lukesteensen — APPROVED** [review]
One small thing to fix that was my fault, but otherwise I think this looks good!

File: src/topology/builder.rs
Comment (jdrouet) [reply src/topology/builder.rs]: I fixed these 😉

**Review by jszwedko — APPROVED** [review]
Looks good, thanks @jdrouet ! I pushed one commit fixing lint.

**Comment by jdrouet** [comment]
Rebase/merge conflicts are making this PR complex to follow. Cherry picking changes to #9051.

**Pull request closed without merging**
